connectors.md

```diff
@@ -1,26 +1,90 @@
 # Connectors and models
 
-## What is a connector?
+This document is intended as a design document for people who want to write connectors for various backends. Its main purpose is to specify what properties connectors should have so they can be easily shared with other people and used in Apollo without any shims.
 
-A connector is the piece of code that links your GraphQL server to different backends. Connectors batch requests to the backend and cache responses. Good connector have two properties:
+This is a draft at the moment, and not the final document. Chances are that the spec will change as we learn about the better ways to build GraphQL servers. It should be pretty close to the final version though, so if you want to get started and build connectors for specific backends, this document is a good starting point.
 
-- Make it as easy as possible to fetch data from different backends
-- Make fetching from backends efficient by batching requests and caching responses
 
-## What is a model?
+Technically you could write a GraphQL server without connectors and models by writing all your logic directly into the resolve functions, but in most cases that's not ideal. Connectors and models are a way of organizing code in a GraphQL server, and you should use them to keep your server modular. If the need arises, you can always write optimized queries directly in your resolvers or models should the need arise.
+
+Let's use an example schema, because it's always easier to explain things with examples:
+```
+type Author {
+  id: ID!
+  name: String
+  posts: [Post]
+}
+
+type Post {
+  id: ID!
+  title: String
+  text: String
+  views: Int
+  author: Author
+}
+
+type Query {
+  author(id: ID!): Author
+  searchPosts(titleContains: String): [Post]
+}
+```
+
+In this example, an author has mutliple posts, and each post has one author.
+
+Here's an illustration for how connectors and models would look like for this example if Authors and Posts were stored in MySQL, but view counts in MongoDB:
+
+![Connectors are database-specfic, models are application-specific](connector-model-diagram.png)
+
+
+The Posts model connects to both SQL and MongoDB. Title, text and authorId come from SQL, the view count comes from MongoDB.
+
+
+## What's in a connector?
+
+A connector is the piece of code that links a GraphQL server to a specific backend (eg. MySQL, MongoDB, S3, neo4j). Each backend will have its own connector. Apart from connecting the GraphQL server to a backend, connectors should also:
+
+- Batch requests together whenever it makes sense
+- Cache data fetched for the backend to avoid extra requests (at least for the duration of one query)
+- Provide a way to log information about data fetched, such as how long the request took, which things were batched together, what was fetched from the cache, how much data was fetched etc.
```

Comment:
I'd add a paragraph below this explaining why this is specific to GraphQL - for example, the GraphQL query structure means batching is very important because a query can fetch the same endpoint many times.